src/network/stream_reader.rs

```diff
@@ -21,33 +21,31 @@
 //!
 
 use std::fmt;
-use std::io;
-use std::io::Read;
+use std::io::{self, BufRead};
 
-use network::message::RawNetworkMessage;
-use consensus::encode;
+use consensus::{encode, Decodable};
 
 /// Struct used to configure stream reader function
-pub struct StreamReader<R> {
+pub struct StreamReader<R: BufRead> {
     /// Stream to read from
-    pub stream: R,
+    pub stream: Box<R>,
```

Comment:
Why do we need a box here?

Reply:
Because R is not Sizable as a trait object (see also my response below)

Reply:
`R` is not a trait object. Just put a `Sized` bound on it if you need a sized bound.

Reply:
Can you pls help me understanding, if we have potentially large object, why do we need to place it inside this struct by value, not by Box or reference? Or you meant that I can put a `Sized` bound and use a reference?

Reply:
Why would you not want it to be by value here?
what's the downside of doing it by value?

Reply:
Probably I was too used over the last 25+ yrs to C++ that I perceive putting large blobs into structs by value as a wrong way of doing things. With variable immutability and copy-on-write it should be a preferable way of building data structures... Thanks for pointing that out.

Reply:
@elichai the problem with having unboxed value is the following: you can't use `StreamReader` to read from the STDIN, since `std::io::StdinLock`, while still implementing `Read` trait, is not `Sized` and can't be embedded into `StreamReader` by value... Any thoughts?

Reply:
Are you sure it doesn't work? 
https://play.rust-lang.org/?gist=385955c4dd46caa4667d8162bc97d0d0

Reply:
That is strange.. My code

```rust

    let mut stream_reader = match filename {
        Some(name) => StreamReader::new(io::BufReader::new(File::open(name)?), None),
        None => {
            let stdin = io::stdin();
            let stream = stdin.lock() as BufRead;
            StreamReader::new(stream, None)
        },
    };
```
using unboxed version of `StreamReader` does not compile with this message:

```
error[E0620]: cast to unsized type: `std::io::StdinLock<'_>` as `dyn std::io::BufRead`
  --> src/main.rs:41:26
   |
41 |             let stream = stdin.lock() as BufRead;
   |                          ^^^^^^^^^^^^^^^^^^^^^^^
   |
help: consider using a box or reference as appropriate
  --> src/main.rs:41:26
   |
41 |             let stream = stdin.lock() as BufRead;
   |                          ^^^^^^^^^^^^

error: aborting due to previous error
```

With boxed version it all worked well. I can't see the difference between your code (that is compiling) and mine though...

Reply:
there is a big difference. You're casting an object into a Trait object.
Why?
trait objects aren't sized, which is why they must be boxed or passed by reference

Reply:
And if you really want to have trait objects then box them yourself and pass the box. Boxs impl DeRef to the underlying type, so that Box actually now implements Read by itself.
https://play.rust-lang.org/?gist=63b487ac5977086f9ec094ad7ed608ab

Reply:
I think I have undertoond my problem. It's not a problem with `StreamReader`, it's just impossible to have same var with either File stream or StdinLock stream type inside b/c they have different types... I need to refactor the code, not `StreamReader`. Sorry for taking your time

Reply:
It's fine :)
Yeah that's something that can be confusing, the difference between generics(like c++ templates) and Trait objects(more like c++ parent classes without the possibility of object slicing(i.e. being unsized)).

Reply:
@elichai I need your help, please: while trying to compile https://github.com/pandoracore/bpdump/blob/master/src/main.rs that uses updated StreamReader like you have shown, I am getting compilation errors like
```
error[E0599]: no method named `read_next` found for type `bitcoin::network::stream_reader::StreamReader<std::boxed::Box<dyn std::io::Read>>` in the current scope
  --> src/main.rs:51:29
   |
51 |         match stream_reader.read_next::<u32>() {
   |                             ^^^^^^^^^
```

while your code with the same modification compiles and works perfectly:
https://play.rust-lang.org/?gist=e5d9473efba6299cea48e7ebcf24fd2d

Reply:
That's because StreamReader exposes a `next_message()` function. StreamReader itself doesn't implement Read.
https://docs.rs/bitcoin/0.19.2/bitcoin/network/stream_reader/struct.StreamReader.html

Reply:
the function was called `next_message` before my last changes; now the name was updated: 
https://github.com/rust-bitcoin/rust-bitcoin/pull/318/files#diff-6e8312b001abf9f9416db1d1ef36830aR58

and it works when I add Box<R> to StreamReader itself. This is very strange...

Reply:
I have an idea of why this could happen, I'll check later when I'm next to a computer

Reply:
Anyway, I spent two days trying to figure out the issue, and from what I see it just seems to be a bug with Cargo getting patch from github branch of top of crates.io and applying it. Basically, rust compiler does not see new function name (it was seeing it before and with Box<R> in rust-bitcoin, but who knows how these compilers do their bugs :)

So I just suggest to review my last commits and, if they are ok, merge them into the master, and compiler bug would not matter.